cassiopeia/simulator/CellSubsampler.py

```diff
@@ -0,0 +1,124 @@
+        the leaves are sampled uniformly at random. Only one of the two
+        criteria can be provided.
+        """
+        if ratio is None and n_cells is None:
+            raise CellSubsamplerError(
+                "At least one of 'ratio' and 'n_cells' " "must be specified."
+            )
+        if ratio is not None and n_cells is not None:
+            raise CellSubsamplerError(
+                "Exactly one of 'ratio' and 'n_cells'" "must be specified."
+            )
+        self.__ratio = ratio
+        self.__n_cells = n_cells
+
+    def subsample(self, tree: CassiopeiaTree) -> CassiopeiaTree:
+        ratio = self.__ratio
+        n_cells = self.__n_cells
+        n_subsample = (
+            n_cells if n_cells is not None else int(tree.n_cell * ratio)
+        )
+        if n_subsample == 0:
+            raise CellSubsamplerError(
+                "ratio too low: no cells would be " "sampled."
+            )
+
+        # First determine which nodes are part of the induced subgraph.
+        leaf_keep_idx = np.random.choice(
+            range(tree.n_cell), n_subsample, replace=False
+        )
+        leaves_in_induced_subtree = [tree.leaves[i] for i in leaf_keep_idx]
+        induced_subtree_degs = dict(
+            [(leaf, 0) for leaf in leaves_in_induced_subtree]
+        )
+
+        nodes_in_induced_subtree = set(leaves_in_induced_subtree)
+        for node in tree.depth_first_traverse_nodes(postorder=True):
+            children = tree.children(node)
+            induced_subtree_deg = sum(
+                [child in nodes_in_induced_subtree for child in children]
+            )
+            if induced_subtree_deg > 0:
+                nodes_in_induced_subtree.add(node)
+                induced_subtree_degs[node] = induced_subtree_deg
+
+        # For debugging:
+        # print(f"leaves_in_induced_subtree = {leaves_in_induced_subtree}")
+        # print(f"nodes_in_induced_subtree = {nodes_in_induced_subtree}")
+        # print(f"induced_subtree_degs = {induced_subtree_degs}")
+        nodes = []
+        edges = []
+        up = {}
+        for node in tree.depth_first_traverse_nodes(postorder=False):
+            if node == tree.root:
+                nodes.append(node)
+                up[node] = node
+                continue
+
+            if node not in nodes_in_induced_subtree:
+                continue
+
+            if induced_subtree_degs[tree.parent(node)] >= 2:
+                up[node] = tree.parent(node)
+            else:
+                up[node] = up[tree.parent(node)]
+
+            if (
+                induced_subtree_degs[node] >= 2
+                or induced_subtree_degs[node] == 0
+            ):
+                nodes.append(node)
+                edges.append((up[node], node))
+        subtree_topology = nx.DiGraph()
+        subtree_topology.add_nodes_from(nodes)
+        subtree_topology.add_edges_from(edges)
+        res = CassiopeiaTree(
+            tree=subtree_topology,
```

Comment:
What about the character matrix that was in the original CassiopeiaTree?

And what about other tree-specific attributes like fitness etc?